### PR TITLE
Adding Feature for Hiding Other Player's Loot

### PR DESCRIPTION
**Pull request by BryanHaley**
This is an implementation of the feature request at #3801 

**What does it do?**
This adds an option to the Ground Items plugin that prevents the overlay from being shown on loot that drops from NPCs killed by other players.
screen shot 2018-11-30 at 12 08 39 am

Here it is in action:
screen shot 2018-11-29 at 10 40 58 pm

**What is the purpose of this feature?**
This is a quality-of-life feature for players in Ironman mode. It makes it much easier to differentiate from NPC drops that can and cannot be picked up.

**How does it work?**
The Ground Items plugin now subscribes to onNpcLootReceived and onPlayerLootReceived events, and keeps track of items dropped by NPCs killed by the player, and the location at which they drop. Then, when it's time for the overlay to be rendered, the item that is being checked for an overlay is compared to the list of loot items, and skipped if it is not present in that list.

Items are removed from the loot item list when they despawn. To avoid memory leaks, the loot item list is also cleared when the game enters the "LOADING" state, as this implies the player is moving away from the area (despawn events are not sent once the player has left the area the item is in, afaik).

**Are there any downsides?**
To my knowledge, there is no way to track specific items on the ground; rather, the best I can do is to track the item type and location. For this reason, this option cannot guarentee 100% accuracy for a variety of reasons, including but not limited to: non-pickup-able items dropping in the same tile as pickup-able items of the same type, picking up and re-dropping items, leaving then returning to the area, et cetera.

In addition, since the overlay checks the loot item list every frame, there can be a significant performance penalty while the option is enabled. My MBP 2015 i5 handles it without any noticeable drop in framerate, but results may vary on weaker machines.

### CODE REVIEW

**Comment by forsco** [comment]
Have you covered all the edge cases for this? Static item spawns? Loot that drops to the ground from interfaces?

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 1]
```diff
@@ -37,277 +37,288 @@
...
 {
 	@ConfigItem(
-		keyName = "highlightedItems",
-		name = "Highlighted Items",
-		description = "Configures specifically highlighted ground items. Format: (item), (item)",
-		position = 0
+			keyName = "highlightedItems",
+			name = "Highlighted Items",
+			description = "Configures specifically highlighted ground items. Format: (item), (item)",
+			position = 0
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 2]
```diff
@@ -37,277 +37,288 @@
...
 		return "";
 	}
 
 	@ConfigItem(
-		keyName = "highlightedItems",
-		name = "",
-		description = ""
+			keyName = "highlightedItems",
+			name = "",
+			description = ""
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 3]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "hiddenItems",
-		name = "Hidden Items",
-		description = "Configures hidden ground items. Format: (item), (item)",
-		position = 1
+			keyName = "hiddenItems",
+			name = "Hidden Items",
+			description = "Configures hidden ground items. Format: (item), (item)",
+			position = 1
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 4]
```diff
@@ -37,277 +37,288 @@
...
 		return "Vial, Ashes, Coins, Bones, Bucket, Jug, Seaweed";
 	}
 
 	@ConfigItem(
-		keyName = "hiddenItems",
-		name = "",
-		description = ""
+			keyName = "hiddenItems",
+			name = "",
+			description = ""
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 5]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "showHighlightedOnly",
-		name = "Show Highlighted items only",
-		description = "Configures whether or not to draw items only on your highlighted list",
-		position = 2
+			keyName = "showHighlightedOnly",
+			name = "Show Highlighted items only",
+			description = "Configures whether or not to draw items only on your highlighted list",
+			position = 2
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 6]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "dontHideUntradeables",
-		name = "Do not hide untradeables",
-		description = "Configures whether or not untradeable items ignore hiding under settings",
-		position = 3
+			keyName = "dontHideUntradeables",
+			name = "Do not hide untradeables",
+			description = "Configures whether or not untradeable items ignore hiding under settings",
+			position = 3
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 7]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "recolorMenuHiddenItems",
-		name = "Recolor Menu Hidden Items",
-		description = "Configures whether or not hidden items in right click menu will be recolored",
-		position = 5
+			keyName = "recolorMenuHiddenItems",
+			name = "Recolor Menu Hidden Items",
+			description = "Configures whether or not hidden items in right click menu will be recolored",
+			position = 6
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 8]
```diff
@@ -37,277 +37,288 @@
...
+
 	@ConfigItem(
-		keyName = "highlightTiles",
-		name = "Highlight Tiles",
-		description = "Configures whether or not to highlight tiles containing ground items",
-		position = 6
+			keyName = "highlightTiles",
+			name = "Highlight Tiles",
+			description = "Configures whether or not to highlight tiles containing ground items",
+			position = 7
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 9]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "notifyHighlightedDrops",
-		name = "Notify for Highlighted drops",
-		description = "Configures whether or not to notify for drops on your highlighted list",
-		position = 7
+			keyName = "notifyHighlightedDrops",
+			name = "Notify for Highlighted drops",
+			description = "Configures whether or not to notify for drops on your highlighted list",
+			position = 8
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 10]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "priceDisplayMode",
-		name = "Price Display Mode",
-		description = "Configures what price types are shown alongside of ground item name",
-		position = 8
+			keyName = "priceDisplayMode",
+			name = "Price Display Mode",
+			description = "Configures what price types are shown alongside of ground item name",
+			position = 9
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 11]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "itemHighlightMode",
-		name = "Item Highlight Mode",
-		description = "Configures how ground items will be highlighted",
-		position = 9
+			keyName = "itemHighlightMode",
+			name = "Item Highlight Mode",
+			description = "Configures how ground items will be highlighted",
+			position = 10
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 12]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "menuHighlightMode",
-		name = "Menu Highlight Mode",
-		description = "Configures what to highlight in right-click menu",
-		position = 10
+			keyName = "menuHighlightMode",
+			name = "Menu Highlight Mode",
+			description = "Configures what to highlight in right-click menu",
+			position = 11
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 13]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "highlightOverValue2",
-		name = "Highlight > Value",
-		description = "Configures highlighted ground items over either GE or HA value",
-		position = 11
+			keyName = "highlightOverValue2",
+			name = "Highlight > Value",
+			description = "Configures highlighted ground items over either GE or HA value",
+			position = 12
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 14]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "hideUnderValue",
-		name = "Hide < Value",
-		description = "Configures hidden ground items under both GE and HA value",
-		position = 12
+			keyName = "hideUnderValue",
+			name = "Hide < Value",
+			description = "Configures hidden ground items under both GE and HA value",
+			position = 13
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 15]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "defaultColor",
-		name = "Default items color",
-		description = "Configures the color for default, non-highlighted items",
-		position = 13
+			keyName = "defaultColor",
+			name = "Default items color",
+			description = "Configures the color for default, non-highlighted items",
+			position = 14
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 16]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "highlightedColor",
-		name = "Highlighted items color",
-		description = "Configures the color for highlighted items",
-		position = 14
+			keyName = "highlightedColor",
+			name = "Highlighted items color",
+			description = "Configures the color for highlighted items",
+			position = 15
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 17]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "hiddenColor",
-		name = "Hidden items color",
-		description = "Configures the color for hidden items in right-click menu and when holding ALT",
-		position = 15
+			keyName = "hiddenColor",
+			name = "Hidden items color",
+			description = "Configures the color for hidden items in right-click menu and when holding ALT",
+			position = 16
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 18]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "lowValueColor",
-		name = "Low value items color",
-		description = "Configures the color for low value items",
-		position = 16
+			keyName = "lowValueColor",
+			name = "Low value items color",
+			description = "Configures the color for low value items",
+			position = 17
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 19]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "lowValuePrice",
-		name = "Low value price",
-		description = "Configures the start price for low value items",
-		position = 17
+			keyName = "lowValuePrice",
+			name = "Low value price",
+			description = "Configures the start price for low value items",
+			position = 18
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 20]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "mediumValueColor",
-		name = "Medium value items color",
-		description = "Configures the color for medium value items",
-		position = 18
+			keyName = "mediumValueColor",
+			name = "Medium value items color",
+			description = "Configures the color for medium value items",
+			position = 19
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 21]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "mediumValuePrice",
-		name = "Medium value price",
-		description = "Configures the start price for medium value items",
-		position = 19
+			keyName = "mediumValuePrice",
+			name = "Medium value price",
+			description = "Configures the start price for medium value items",
+			position = 20
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 22]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "highValueColor",
-		name = "High value items color",
-		description = "Configures the color for high value items",
-		position = 20
+			keyName = "highValueColor",
+			name = "High value items color",
+			description = "Configures the color for high value items",
+			position = 21
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 23]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "highValuePrice",
-		name = "High value price",
-		description = "Configures the start price for high value items",
-		position = 21
+			keyName = "highValuePrice",
+			name = "High value price",
+			description = "Configures the start price for high value items",
+			position = 22
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 24]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "insaneValueColor",
-		name = "Insane value items color",
-		description = "Configures the color for insane value items",
-		position = 22
+			keyName = "insaneValueColor",
+			name = "Insane value items color",
+			description = "Configures the color for insane value items",
+			position = 23
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java [hunk 25]
```diff
@@ -37,277 +37,288 @@
...
 
 	@ConfigItem(
-		keyName = "insaneValuePrice",
-		name = "Insane value price",
-		description = "Configures the start price for insane value items",
-		position = 23
+			keyName = "insaneValuePrice",
+			name = "Insane value price",
+			description = "Configures the start price for insane value items",
+			position = 24
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsConfig.java]: Revert this diff for the indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 26]
```diff
@@ -29,34 +29,23 @@
...
 import java.awt.Color;
 import java.awt.Rectangle;
 import static java.lang.Boolean.TRUE;
-import java.util.HashSet;
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+
+import java.util.*;
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Don't use wildcard imports

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 27]
```diff
@@ -29,34 +29,23 @@
...
-import net.runelite.api.ItemComposition;
-import net.runelite.api.ItemID;
-import net.runelite.api.ItemLayer;
-import net.runelite.api.MenuAction;
-import net.runelite.api.MenuEntry;
-import net.runelite.api.Node;
-import net.runelite.api.Player;
-import net.runelite.api.Scene;
-import net.runelite.api.Tile;
+import net.runelite.api.*;
```
Comment (forsco) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: Don't use wildcard imports

**Comment by BryanHaley** [comment]
Pushed a commit, please let me know if any further changes are necessary.

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java [hunk 28]
```diff
@@ -261,6 +295,44 @@ public void onItemQuantityChanged(ItemQuantityChanged itemQuantityChanged)
...
+		final Player player = playerLootReceived.getPlayer();
+		final Collection<ItemStack> items = playerLootReceived.getItems();
+
+		for (ItemStack i: items)
+		{
+			lootItems.put(i.getId(), player.getLocalLocation());
+		}
+	}
+
+	boolean isPlayerOwned(GroundItem groundItem)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItemsPlugin.java]: This is too expensive to be called per item per frame. You should probably figure out how to mark `GroundItem`s instead.

File: runelite-api/src/main/java/net/runelite/api/overlay/OverlayIndex.java [hunk 29]
```diff
@@ -52,6 +52,10 @@
 		{
 			log.warn("unable to load overlay index", ex);
 		}
+		catch (NullPointerException ex)
```
Comment (Adam-) [inline runelite-api/src/main/java/net/runelite/api/overlay/OverlayIndex.java]: This is debate-ably incorrect but regardless does not belong in this PR.

File: runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItem.java [hunk 30]
```diff
@@ -38,12 +39,18 @@
 	private String name;
 	private int quantity;
 	private WorldPoint location;
+	private LocalPoint localLoc;
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grounditems/GroundItem.java]: Ground items are cleared by the plugin whenever a new scene loads, so there is never a reason to store both the world and local location.

**Comment by BryanHaley** [comment]
It should, yes.
